pkg/kubelet/server/server_test.go

```diff
@@ -855,6 +863,216 @@ func TestContainerLogsWithInvalidTail(t *testing.T) {
+				{stream: v1.LogStreamTypeStdout, msg: "out1\n"},
+			},
+			query: "?stream=stderr",
+			expectedLogOptions: &v1.PodLogOptions{
+				Stream: &streamStderr,
+			},
+			expectedOutput: "err1\nerr2\n",
+		},
+		{
+			name: "return all logs",
+			logs: []logEntry{
+				{stream: v1.LogStreamTypeStdout, msg: "out1\n"},
+				{stream: v1.LogStreamTypeStderr, msg: "err1\n"},
+				{stream: v1.LogStreamTypeStdout, msg: "out2\n"},
+			},
+			query: "?stream=All",
+			expectedLogOptions: &v1.PodLogOptions{
+				Stream: &streamAll,
+			},
+			expectedOutput: "out1\nerr1\nout2\n",
+		},
+		{
+			name: "stdout logs with legacy tail",
+			logs: []logEntry{
+				{stream: v1.LogStreamTypeStdout, msg: "out1\n"},
+				{stream: v1.LogStreamTypeStderr, msg: "err1\n"},
+				{stream: v1.LogStreamTypeStdout, msg: "out2\n"},
+			},
+			query: "?stream=All&tail=1",
+			expectedLogOptions: &v1.PodLogOptions{
+				Stream:    &streamAll,
+				TailLines: pointer.Int64(1),
+			},
+			expectedOutput: "out2\n",
+		},
+		{
+			name: "last line of logs does not match the stream filter",
+			logs: []logEntry{
+				{stream: v1.LogStreamTypeStdout, msg: "out1\n"},
+				{stream: v1.LogStreamTypeStdout, msg: "out2\n"},
+				{stream: v1.LogStreamTypeStderr, msg: "err1\n"},
+			},
+			query: "?stream=stdout&tailLines=1",
+			expectedLogOptions: &v1.PodLogOptions{
+				Stream:    &streamStdout,
+				TailLines: pointer.Int64(1),
+			},
+			expectedOutput: "",
+		},
+		{
+			name: "last line of logs matches the stream filter",
+			logs: []logEntry{
+				{stream: v1.LogStreamTypeStdout, msg: "out1\n"},
+				{stream: v1.LogStreamTypeStderr, msg: "err1\n"},
+				{stream: v1.LogStreamTypeStdout, msg: "out2\n"},
+			},
+			query: "?stream=stdout&tailLines=1",
+			expectedLogOptions: &v1.PodLogOptions{
+				Stream:    &streamStdout,
+				TailLines: pointer.Int64(1),
+			},
+			expectedOutput: "out2\n",
+		},
+		{
+			name: "return the logs within the last 2 lines that matches the stream filter",
+			logs: []logEntry{
+				{stream: v1.LogStreamTypeStdout, msg: "out1\n"},
+				{stream: v1.LogStreamTypeStderr, msg: "err1\n"},
+				{stream: v1.LogStreamTypeStdout, msg: "out2\n"},
+			},
+			query: "?stream=stderr&tailLines=2",
+			expectedLogOptions: &v1.PodLogOptions{
+				Stream:    &streamStderr,
+				TailLines: pointer.Int64(2),
+			},
+			expectedOutput: "err1\n",
+		},
+		{
+			name: "return the first 6 bytes of the logs within last 3 lines that matches the stream filter",
+			logs: []logEntry{
+				{stream: v1.LogStreamTypeStderr, msg: "err1\n"},
+				{stream: v1.LogStreamTypeStdout, msg: "out1\n"},
+				{stream: v1.LogStreamTypeStderr, msg: "err2\n"},
+				{stream: v1.LogStreamTypeStdout, msg: "out2\n"},
+			},
+			query: "?stream=stdout&tailLines=3&limitBytes=6",
+			expectedLogOptions: &v1.PodLogOptions{
+				Stream:     &streamStdout,
+				TailLines:  pointer.Int64(3),
+				LimitBytes: pointer.Int64(6),
+			},
+			expectedOutput: "out1\no",
+		},
+	}
```

Comment:
> Ideally add a test that those log options will be respected alongside with setting the stream option

@SergeyKanzhelev I have added a corresponding case to demonstrate how `tailLines` and `limitBytes` work with the `stream` option.

BTW I force-pushed so as to keep the git commit history clean. If that is not conveninet for review, I'm okay with pushing only new commits.